kratos/processes/apply_periodic_boundary_condition_process.h

```diff
@@ -0,0 +1,198 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#ifndef APPLY_PERIODIC_CONDITION_PROCESS_H
+#define APPLY_PERIODIC_CONDITION_PROCESS_H
+
+// System includes
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "geometries/point.h"
+#include "processes/process.h"
+#include "utilities/math_utils.h"
+#include "includes/kratos_parameters.h"
+#include "includes/dof.h"
+#include "includes/model_part.h"
+
+namespace Kratos
+{
+
+class KRATOS_API(KRATOS_CORE) ApplyPeriodicConditionProcess : public Process
```

Comment:
For now "ApplyPeriodicConditionProcess" handles matching meshes in serial with translated or rotated boundaries (which is already a lot for one class). How will this be extended to include non-matching periodic bcs and mpi? We don't need a complete design here, but this class (it's name and parameters) should be chosen in such a way that we can extend the periodic conditions in the future without changing it's user interface (open to extension, but closed to modification is what I'm pushing for here). This will pay off in terms of backwards compatibility and duplicating effort by redesigning/rewriting code if it is tackled here. A bonus would be if the implementation of this class could already be designed with these future functionalities in mind.